File: Backend/app/Services/AuthService.php
<?php
namespace App\Services;

use Firebase\JWT\JWT;
use Firebase\JWT\Key;
use App\Models\User;

class AuthService {

    private string $secret;
    private int $ttl = 3600 * 24; // 24 hours

    public function __construct() {
        $this->secret = $_ENV['JWT_SECRET'];
    }

    public function login(string $email, string $password): array {
        $user = User::where('email', $email)->first();

        if (!$user) {
            throw new \Exception('Email ou mot de passe incorrect', 401);
        }

        if ($user->password !== $password) {
            throw new \Exception('Email ou mot de passe incorrect', 401);
        }

        if (!$user->active) {
            throw new \Exception('Compte désactivé', 403);
        }

        $payload = [
            'iss' => 'budgetcollab',
            'sub' => $user->id,
            'role' => $user->role,
            'iat' => time(),
            'exp' => time() + $this->ttl,
        ];

        $token = JWT::encode($payload, $this->secret, 'HS256');

        return [
            'token' => $token,
            'user'  => [
                'id'            => $user->id,
                'name'          => $user->name,
                'email'         => $user->email,
                'role'          => $user->role,
                'active'        => (bool) $user->active,
                'deleteRequest' => (bool) $user->delete_request,
                'createdAt'     => $user->created_at,
                'updatedAt'     => $user->updated_at,
            ]
        ];
    }

    public function validateToken(string $token): object {
        return JWT::decode($token, new Key($this->secret, 'HS256'));
    }
}

File: Backend/app/Services/auth_service.py
from datetime import datetime, timedelta, timezone

import bcrypt
from jose import JWTError, jwt

from app.config import settings


class AuthService:
    @staticmethod
    def hash_password(password: str) -> str:
        return bcrypt.hashpw(password.encode("utf-8"), bcrypt.gensalt()).decode("utf-8")

    @staticmethod
    def verify_password(plain_password: str, hashed_password: str) -> bool:
        return bcrypt.checkpw(plain_password.encode("utf-8"), hashed_password.encode("utf-8"))

    @staticmethod
    def create_token(user_id: str, role: str) -> str:
        expire = datetime.now(timezone.utc) + timedelta(hours=settings.JWT_EXPIRATION_HOURS)
        payload = {
            "sub": user_id,
            "role": role,
            "exp": expire,
            "iat": datetime.now(timezone.utc),
        }
        return jwt.encode(payload, settings.JWT_SECRET, algorithm=settings.JWT_ALGORITHM)

    @staticmethod
    def decode_token(token: str) -> dict | None:
        try:
            payload = jwt.decode(token, settings.JWT_SECRET, algorithms=[settings.JWT_ALGORITHM])
            return payload
        except JWTError:
            return None
